src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+    fn read(&self) -> Result<Vec<Message>, anyhow::Error> {
+        read_segment(&self.path)
+    }
+
+    /// Physically concatenate all batches together into a single batch
+    /// and consolidate updates up to the frontier if provided.
+    fn compact(
+        batches: &[Batch],
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut messages = vec![];
+
+        for batch in batches {
+            messages.append(&mut read_segment(&batch.path)?);
+        }
+
+        Batch::create_from_messages(messages, trace_path, compaction_frontier)
+    }
+}
+
+/// A Trace is an on-disk representation of data meant to mimic a differential Trace.
+///
+/// A Trace listens to the `wal_path` and checks for WAL segments that it can consolidate
+/// into Batches (stored in the `trace_path`). Once it exceeds a certain number of Batches
+/// it tries to physically and logically compact them into a single batch that is
+/// compacted up to the `compaction` frontier.
+#[derive(Debug)]
+pub struct Trace {
+    trace_path: PathBuf,
+    wal_path: PathBuf,
+    batches: Vec<Batch>,
+    compaction: Option<Timestamp>,
+}
+
+impl Trace {
+    fn create(id: GlobalId, trace_path: PathBuf, wal_path: PathBuf) -> Result<Self, anyhow::Error> {
+        let _ = fs::read_dir(&wal_path).with_context(|| {
+            format!(
+                "trying to ensure wal directory {} exists for trace of relation {}",
+                id,
+                wal_path.display()
+            )
+        })?;
+
+        // Create a new directory to store the trace
+        fs::create_dir(&trace_path).with_context(|| {
+            format!("trying to create trace directory: {}", trace_path.display())
+        })?;
+
+        Ok(Self {
+            trace_path,
+            wal_path,
+            batches: Vec::new(),
+            compaction: None,
+        })
+    }
+
+    /// Remove all on-disk data for this trace.
+    ///
+    /// Importantly, we also delete the WAL directory here (the WAL writer
+    /// only gets to add new files and can't do anything else)
+    fn destroy(self) -> Result<(), anyhow::Error> {
+        fs::remove_dir_all(&self.trace_path).with_context(|| {
+            format!(
+                "failed to remove trace directory {}",
+                self.trace_path.display()
+            )
+        })?;
+        fs::remove_dir_all(&self.wal_path).with_context(|| {
+            format!("failed to remove wal directory {}", self.wal_path.display())
+        })?;
+
+        Ok(())
+    }
+
+    /// Checks if there are finished WAL segments and if so, forms them into batches.
+    fn consume_wal(&mut self) -> Result<(), anyhow::Error> {
+        let finished_segments = self.find_finished_wal_segments()?;
+
+        for segment in finished_segments {
+            let batch = Batch::create(&segment, &self.trace_path)?;
+            self.batches.push(batch);
```

Comment:
Going through the code it's difficult for me to assert that batches are always in sorted order. For example, is `consume_wal` always called after `find_batches` makes its way into the Vec?

Would it make sense to make this a BTreeMap and not have to worry about keeping the order manually?